### PR TITLE
Add Containers HTTP/GRPC interface to list active containers by workspace

### PR DESCRIPTION
**Pull request by jsun-m**
- Add a schedulerContainerWorkspaceIndex Redis key during container scheduling. This key should serve as a Redis index, allowing the container state to be queried by workspaceId.
- Add HTTP endpoint for querying containers by workspace
- Add GRPC service for querying containers by workspace
- Added CLI command to query and render containers

### CODE REVIEW

**Comment by jsun-m** [comment]
90% of the changes here are from protobuf codegen

File: pkg/api/v1/container.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, "Failed to get containers")
+	}
+
+	return ctx.JSON(http.StatusOK, containerStates)
+}
+
+func (g *ContainerGroup) authorize(ctx echo.Context) (*auth.HttpAuthContext, error) {
+	cc, _ := ctx.(*auth.HttpAuthContext)
+	if cc.AuthInfo.Token.TokenType != types.TokenTypeClusterAdmin {
```
Comment (luke-lombardi) [inline pkg/api/v1/container.go:58]: I think this check should be on the workspace ID == token workspace ID, so users can list their containers

File: pkg/api/v1/stub.go [hunk 2]
```diff
@@ -38,6 +39,7 @@ func (g *StubGroup) ListStubs(ctx echo.Context) error {
 	}
 
 	if stubs, err := g.backendRepo.ListStubs(ctx.Request().Context(), filters); err != nil {
+		log.Println(err)
```
Comment (luke-lombardi) [inline pkg/api/v1/stub.go:42]: remove log

File: pkg/repository/backend_postgres.go [hunk 3]
```diff
@@ -885,14 +884,35 @@ func (c *PostgresBackendRepository) CreateDeployment(ctx context.Context, worksp
...
 
+func (c *PostgresBackendRepository) listStubsQueryBuilder(filters types.StubFilter) squirrel.SelectBuilder {
+	qb := squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar).Select(
+		"s.*",
+		"w.external_id AS \"workspace.external_id\"", "w.name AS \"workspace.name\"",
+	).From("stub s").
+		Join("workspace w ON s.workspace_id = w.id")
+
+	// Apply filters
+	if filters.WorkspaceID != "" {
```
Comment (luke-lombardi) [inline pkg/repository/backend_postgres.go:895]: do we still need this?

File: pkg/repository/container_redis.go [hunk 4]
```diff
@@ -73,6 +80,13 @@ func (cr *ContainerRedisRepository) SetContainerState(containerId string, info *
 		return fmt.Errorf("failed to add container state key to index <%v>: %w", indexKey, err)
 	}
 
+	// Add another state key to workspace index (by stub id)
```
Comment (luke-lombardi) [inline pkg/repository/container_redis.go:83]: I think this comment should be updated

File: pkg/repository/container_redis.go [hunk 5]
```diff
@@ -201,16 +215,9 @@ func (cr *ContainerRedisRepository) GetWorkerAddress(containerId string) (string
...
 
-// GetActiveContainersByStubId gets active containers by stub ID.
-// stubId is the external ID.
-func (cr *ContainerRedisRepository) GetActiveContainersByStubId(stubId string) ([]types.ContainerState, error) {
-	indexKey := common.RedisKeys.SchedulerContainerIndex(stubId)
-	keys, err := cr.rdb.SMembers(context.TODO(), indexKey).Result()
-	if err != nil {
-		return nil, fmt.Errorf("failed to retrieve container state keys: %v", err)
-	}
+func (cr *ContainerRedisRepository) listContainerKeysByIndexKeys(indexKey string, keys []string) ([]types.ContainerState, error) {
```
Comment (luke-lombardi) [inline pkg/repository/container_redis.go:218]: rename -> listContainerStateByIndex or something like that and add a comment about what the func does

**Review by luke-lombardi — CHANGES_REQUESTED** [review]

File: pkg/repository/backend_postgres.go (hunk 3 above)
Comment (jsun-m) [reply pkg/repository/backend_postgres.go:895]: it's good to include. I need to update the http endpoint for stubs later to include workspace and workspace authentication

File: pkg/repository/container_redis.go (hunk 5 above)
Comment (jsun-m) [reply pkg/repository/container_redis.go:218]: updated

**Review by luke-lombardi — APPROVED** [review]
LGTM

**Pull request merged**